### PR TITLE
[DISCUSSION] Add hash_object constraint to causal value hash.

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

The trunk version of the schema allows causals to reference branch hashes which might not have a corresponding branch object, but UCM still currently expects all branch objects to exist, we'd discussed allowing the possibility of a "shallow" causal spine push/pull which only pushes the hashes of the history but not all the branches (this would be faster and more space efficient), but for now we don't have that implemented.

This change adds a constraint to ensure this assumption isn't violated again, since its violation caused a couple problems on the last migration.

This PR is a just a discussion, since this change ended up being much more invasive than I originally anticipated, so I'm not yet convinced that the complications it introduces are worth the value it brings. Let's chat about this in a meeting before moving forwards.

* Keeps separate v*.sql version files for each schema version for historical reasons
* Adds a migration which swaps the `hash(id)` reference constraint to `hash_object(hash_id)` to ensure value_hash's have a corresponding object.
  * It turns out that sqlite3 doesn't actually allow changing constraints in-place on a table! You have to create a brand new column/table to do this 😬 , not actually that hard to do, but definitely annoying.
* Adds this constraint to the latest `v3.sql` schema creation code.
* Rewrites the `saveCausal` codepath to save things in the order required by these constraints.

## Implementation notes

Adds a step to the v3 migration which re-writes the `causal` table to include the new `hash_object` constraint.

### CODE REVIEW

File: codebase2/codebase-sqlite/sql/schemas/v3.sql [hunk 1]
```diff
@@ -0,0 +1,240 @@
+-- v3 codebase schema
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/sql/schemas/v3.sql:1]: Just a copy of v2, but with the new constraint on `TABLE causal`'s value hash.

File: codebase2/codebase-sqlite/sql/schemas/v3.sql [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+-- look up objects by primary_hash_id.
+CREATE UNIQUE INDEX object_hash_id ON object(primary_hash_id);
+-- filter objects by
+CREATE INDEX object_type_id ON object(type_id);
+
+CREATE TABLE causal (
+  -- The hash representing a branch and its history
+  self_hash_id INTEGER PRIMARY KEY NOT NULL CONSTRAINT causal_fk1 REFERENCES hash(id),
+  -- The hash representing just the branch at this point in time, no history.
+  value_hash_id INTEGER NOT NULL CONSTRAINT causal_fk2 REFERENCES hash_object(hash_id)
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/sql/schemas/v3.sql:91]: This line has the only new change (and I added comments to the fields)

File: codebase2/codebase-sqlite/sql/schemas/v1.sql [hunk 3]
```diff
@@ -0,0 +1,241 @@
+-- v1 codebase schema
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/sql/schemas/v1.sql:1]: The same as the v2 schema, but with `1` as the schema version. I added this for historical record.

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs [hunk 4]
```diff
@@ -880,6 +882,24 @@ removeHashObjectsByHashingVersion hashVersion =
       WHERE hash_version = ?
 |]
 
+addHashObjectConstraintToCausalTable :: DB m => m ()
+addHashObjectConstraintToCausalTable =
+  execute_ sql
+  where
+    sql =
+      [here|
+-- Not sure why, but we need this savepoint or this operation fails.
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:891]: I need to explore this a bit further, without this savepoint we get a "TABLE causal does not exist" error later on...

**Comment by aryairani** [comment]
How invasive would it be to add the branch `object_id` as an ignored parameter to the Haskell function that creates the `causal` entries; adding the constraint in Haskell instead of SQL?

**Comment by ChrisPenner** [comment]
We discussed it, we'll probably be adding support for "shallow" histories soon after we have a github replacement, so this likely isn't worth the effort at the moment since it would just need to be undone at that time 👍🏼

**Pull request closed without merging**